src/persist/src/storage.rs

```diff
@@ -211,6 +211,36 @@ pub trait Blob: BlobRead + Sized {
     async fn delete(&mut self, key: &str) -> Result<(), Error>;
 }
 
+/// An abstraction for a location in storage that supports an arbitrary blob of binary data
+/// and a sequence number, where the data and sequence number are conditionally
+/// updated.
+///
+/// Users are expected to use this API with consistently increasing sequence numbers
+/// to allow multiple processes across multiple machines to agree to a total order
+/// of the evolution of this binary data.
+#[async_trait]
+pub trait Consensus: Sized {
+    /// The configuration necessary to open this type of storage.
+    type Config;
+
+    /// Opens the given location.
+    fn open(config: Self::Config) -> Result<Self, Error>;
+
+    /// Returns a recent version of `data`, and the corresponding sequence number, if
+    /// one exists at this location.
+    async fn head(&self) -> Result<Option<(SeqNo, Vec<u8>)>, Error>;
+
+    /// Update the data and sequence number stored at this location to `new`,
+    /// iff the current sequence number is exactly `current`.
+    ///
+    /// Returns recent value of seqno iff the current value does not equal `current`.
+    async fn cas(
+        &mut self,
+        current: Option<SeqNo>,
+        new: Option<(SeqNo, Vec<u8>)>,
```

Comment:
why is this an option? to allow for eventually deleting it if we delete the shard and leave only a tombstone in Blob storage?

Reply:
yeah

Reply:
okay a question I think we should think about then is e.g. is it allowed to set this to None and then bring it back? if not, then should it be modeled more like an explicit delete? (these are actual questions, I don't have an opinion here yet)

Reply:
I didn't add an explicit comment on this -- my reaction after thinking about it is that I think ... that's fine? It seems to me to be similar to the case where we go from seqno A to seqno B, and then back from seqno B to seqno A - its weird but its not disallowed by the api. The other way to do a delete in this world would be to set the data to the empty vec?

Reply:
but i think this is a question that (as you said) is best left open while we continue to sketch out a prototype and think about it